src/vmm/src/devices/virtio/net/device.rs

```diff
@@ -102,6 +89,128 @@ pub struct ConfigSpace {
+    /// `DescriptorChain`. (See also related comment in
+    /// [`IoVecBufferMut::append_descriptor_chain`]).
+    unsafe fn add_buffer(
+        &mut self,
+        mem: &GuestMemoryMmap,
+        head: DescriptorChain,
+    ) -> Result<(), AddRxBufferError> {
+        let parsed_dc = self.iovec.append_descriptor_chain(mem, head)?;
+        if parsed_dc.length < self.min_buffer_size {
+            self.iovec.drop_descriptor_chain(&parsed_dc);
+            return Err(AddRxBufferError::BufferTooSmall);
+        }
+        self.parsed_descriptors.push_back(parsed_dc);
+        Ok(())
+    }
+
+    /// Returns the number of available `iovec` objects.
+    #[inline(always)]
+    fn len(&self) -> usize {
+        self.iovec.len()
+    }
+
+    /// Mark the first `size` bytes of available memory as used.
+    ///
+    /// The `size` should be the number of bytes previously written in the [`RxBuffers`]. This
+    /// means that `size` cannot be more than the currently available bytes in the underlying
+    /// `IoVecBufferMut`.
+    fn mark_used(&mut self, size: u32, metrics: &NetDeviceMetrics) {
+        // Since we were able to write a frame in guest memory, we should have at least one
+        // descriptor chain here. If not, we have a bug, so fail fast, since the device is
+        // fundamentally broken.
+        let mut parsed_dc = self.parsed_descriptors.pop_front().expect(
+            "net: internal bug. Mismatch between written frame size and available descriptors",
+        );
+
+        // We are trying to mark as used more bytes than a single descriptor. However without
+        // VIRTIO_NET_F_MRG_RXBUF negotiated (we don't support it currently) the guest can't handle
+        // frames spanning multiple DescriptorChains. There is nothing we can do now, just pass the
+        // descriptor back to the guest without claiming any written bytes
+        if size > parsed_dc.length {
+            parsed_dc.length = 0;
+            metrics.rx_fails.inc();
+        } else {
+            parsed_dc.length = size;
+            metrics.rx_bytes_count.add(size as u64);
+            metrics.rx_packets_count.inc();
+        }
+        self.header_set_num_buffers(1);
+        self.iovec.drop_descriptor_chain(&parsed_dc);
+        self.deferred_descriptor = Some(parsed_dc);
+    }
+
+    /// Write the number of descriptors used in VirtIO header
+    fn header_set_num_buffers(&mut self, nr_descs: u16) {
+        // We can unwrap here, because we have checked before that the `IoVecBufferMut` holds at
+        // least one buffer with the proper size, depending on the feature negotiation. In any
+        // case, the buffer holds memory of at least `std::mem::size_of::<virtio_net_hdr_v1>()`
+        // bytes.
+        self.iovec
+            .write_all_volatile_at(
+                &nr_descs.to_le_bytes(),
+                std::mem::offset_of!(virtio_net_hdr_v1, num_buffers),
+            )
```

Comment:
If that is true we should add a check that the first descriptor is at least size_of(virtio_net_hdr_v1) in the emulation code. Then, this will be true. However, I didn't find in the spec this requirement. Could you point me to it?

Reply:
Ok, is seems I am incorrect here. I was thinking the size requirements were for actual buffers descriptors point to (so the minimum is size is for the header alone), but it seems it refers to the whole descriptor chain capacity.